presto-main/src/main/java/com/facebook/presto/server/remotetask/WorkerTaskStatusFetcher.java

```diff
@@ -0,0 +1,314 @@
+
+        this.onFail = requireNonNull(onFail, "onFail is null");
+
+        this.refreshMaxWait = requireNonNull(refreshMaxWait, "refreshMaxWait is null");
+        this.taskListStatusCodec = requireNonNull(taskListStatusCodec, "taskStatusCodec is null");
+
+        this.executor = requireNonNull(executor, "executor is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+
+        this.errorTracker = new BatchRequestErrorTracker(
+                Collections.list(idTaskMap.keys()),
+                getWorkerURI(),
+                maxErrorDuration,
+                errorScheduledExecutor,
+                "getting task status");
+        this.stats = requireNonNull(stats, "stats is null");
+        this.isBinaryTransportEnabled = isBinaryTransportEnabled;
+
+        this.sessionId = sessionId;
+    }
+
+    public void addTask(ContinuousBatchTaskStatusFetcher.Task task)
+    {
+        idTaskMap.put(
+                task.taskId,
+                new Pair<ContinuousBatchTaskStatusFetcher.Task, StateMachine<TaskStatus>>(
+                        task,
+                        new StateMachine(
+                                "task-" + task.taskId, executor, task.taskStatus)));
+    }
+
+    TaskStatus getTaskStatus(TaskId taskId)
+    {
+        return idTaskMap.get(taskId).getKey().taskStatus.get();
+    }
+
+    private URI getWorkerURI()
+    {
+        URI uri;
+        try {
+            uri = uriBuilderFrom(new URI(worker)).appendPath("v1/task/status?session_id=" + sessionId).build();
+        }
+        catch (URISyntaxException e) {
+            uri = null;
+        }
+        return uri;
+    }
+
+    public synchronized void scheduleNextRequest()
+    {
+        // stopped or done?
+        if (!running) {
+            return;
+        }
+
+        // outstanding request?
+        if (future != null && !future.isDone()) {
+            // this should never happen
+            log.error("Can not reschedule update because an update is already running");
+            return;
+        }
+
+        // if throttled due to error, asynchronously wait for timeout and try again
+        ListenableFuture<?> errorRateLimit = errorTracker.acquireRequestPermit();
+        if (!errorRateLimit.isDone()) {
+            errorRateLimit.addListener(this::scheduleNextRequest, executor);
+            return;
+        }
+
+        HashMap<TaskId, TaskState> idStateMap = new HashMap<>(); // Maybe we need to make this JSON instead?
+        for (Pair<ContinuousBatchTaskStatusFetcher.Task, StateMachine<TaskStatus>> taskPair : idTaskMap.values()) {
+            ContinuousBatchTaskStatusFetcher.Task task = taskPair.getKey();
+            idStateMap.put(task.taskId, task.taskStatus.get().getState());
+        }
+
+        Request request = setContentTypeHeaders(isBinaryTransportEnabled, prepareGet())
+                .setUri(getWorkerURI())
+                .setHeader(PRESTO_CURRENT_STATE, idStateMap.toString())
```

Comment:
not needed as we are not passing any existing state to worker. Worker is keeping track of the previous state using session id.